### PR TITLE
Support use alias column in indices

### LINKED ISSUE #15449: ALIAS column in secondary index usage.
**Use case**
We have some complex function, which we would like to use in secondary index, and we want to create an alias column as that function in order to have simple and memoriable name to it, which we can use in WHERE clause.

**Describe the solution you'd like**
Clickhouse server version 20.9.2.20
```
CREATE TABLE test_index
(
    `key_string` String,
    `key_uint32` ALIAS toUInt32(key_string),
    INDEX idx key_uint32 TYPE set(0) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY tuple()
PRIMARY KEY tuple()
ORDER BY key_string


Received exception from server (version 20.9.2):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'key_uint32' while processing query: 'key_uint32', required columns: 'key_uint32', source columns: 'key_string'.


CREATE TABLE test_index
(
    `key_string` String,
    `key_uint32` ALIAS toUInt32(key_string),
    INDEX idx toUInt32(key_string) TYPE set(0) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY tuple()
PRIMARY KEY tuple()
ORDER BY key_string;

INSERT INTO test_index SELECT * FROM numbers(1000000);

SELECT *
FROM test_index
WHERE key_uint32 = 1

┌─key_string─┐
│ 1          │
└────────────┘

1 rows in set. Elapsed: 0.010 sec. Processed 1.00 million rows, 14.89 MB (99.93 million rows/s., 1.49 GB/s.)

SELECT *
FROM test_index
WHERE toUInt32(key_string) = 1

┌─key_string─┐
│ 1          │
└────────────┘

1 rows in set. Elapsed: 0.008 sec. Processed 8.19 thousand rows, 121.96 KB (1.02 million rows/s., 15.20 MB/s.)
```
**Describe alternatives you've considered**
Write that kind of conditions in full form, but that makes less sense when we have the ability to use ALIAS columns.

### LINKED ISSUE #55650: Cannot use ALIAS columns in indices
This does not work:
```
CREATE TABLE test
(
    x UInt32,
    y ALIAS x + 1,
    INDEX i_y (y) TYPE minmax
) ENGINE = MergeTree ORDER BY x;
```

This does:
```
CREATE TABLE test
(
    x UInt32,
    y ALIAS x + 1,
    INDEX i_y (x + 1) TYPE minmax
) ENGINE = MergeTree ORDER BY x;
```

### PR DESCRIPTION
**Pull request by ucasfl**
Fixes #15449 
Fixes #55650

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
It is now possible to refer to ALIAS column in index (non-primary-key) definitions (issue #55650). Example: `CREATE TABLE tab(col UInt32, col_alias ALIAS col + 1, INDEX idx (col_alias) TYPE minmax) ENGINE = MergeTree ORDER BY col;`

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit e603874783d1313a0577653987748fc87f63c3bc with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57220/e603874783d1313a0577653987748fc87f63c3bc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6989892186">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57220/e603874783d1313a0577653987748fc87f63c3bc/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/ReplaceAliasToExprVisitor.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#include <Interpreters/InDepthNodeVisitor.h>
+
+namespace DB
+{
+
+class ASTFunction;
+class ColumnsDescription;
+class ASTIdentifier;
+
+class ReplaceAliasToExprMatcher
```
Comment (rschu1ze) [inline src/Storages/ReplaceAliasToExprVisitor.h:13]: Please add a class-level comment what the visitor does and in which situations it can be used. Alternatively, please move the visitor code into an anonymous namespace or as static functions into IndicesDescription.cpp as this is the only place where it is used (and I think it will stay that way).

File: src/Storages/IndicesDescription.cpp [hunk 2]
```diff
@@ -94,6 +96,10 @@ IndexDescription IndexDescription::getIndexFromAST(const ASTPtr & definition_ast
     if (index_definition->expr)
     {
         expr_list = extractKeyExpressionList(index_definition->expr->clone());
+
+        ReplaceAliasToExprVisitor::Data data{columns};
```
Comment (rschu1ze) [inline src/Storages/IndicesDescription.cpp:100]: Cosmetic: `ReplaceAliasByExpressionVisitor`

File: src/Storages/ReplaceAliasToExprVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTLiteral.h>
+#include <Storages/ColumnsDescription.h>
+#include <Storages/ReplaceAliasToExprVisitor.h>
```
Comment (rschu1ze) [inline src/Storages/ReplaceAliasToExprVisitor.cpp:4]: The corresponding header file should be the first include.

File: tests/queries/0_stateless/02911_support_alias_column_in_indices.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    y1 ALIAS x + 1,
+    y2 ALIAS y1 + 1,
+    INDEX i_y (y2) TYPE minmax
+) ENGINE = MergeTree ORDER BY tuple();
+
+insert into test1 select * from numbers(10);
+insert into test1 select * from numbers(11, 20);
+
+explain indexes = 1 select * from test where y > 10;
+explain indexes = 1 select * from test1 where y2 > 15;
```
Comment (ucasfl) [inline tests/queries/0_stateless/02911_support_alias_column_in_indices.sql:29]: Analyzer does not support this case of index, index does not hit for it.

Comment (rschu1ze) [reply tests/queries/0_stateless/02911_support_alias_column_in_indices.sql]: As far as I see, the EXPLAIN PLAN output is slightly different for analyzer. I added a separate test.

Comment (rschu1ze) [reply tests/queries/0_stateless/02911_support_alias_column_in_indices.sql]: After more testing, I now see that the analyzer does not pick up the skipping index if it contains a "nested" alias. I think this is more an edge case and it is possible to work around it, so I would be fine to merge this PR for now and create a follow-up issue.

File: tests/queries/0_stateless/02911_support_alias_column_in_indices.sql [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    a1 alias c + 1,
+    a2 alias a1 + 1,
+    index i (a2) type minmax
+) engine = MergeTree order by c;
+
+insert into test2 select * from numbers(10);
+insert into test2 select * from numbers(11, 20);
+
+explain indexes = 1 select * from test2 where a2 > 15 settings allow_experimental_analyzer = 0;
+explain indexes = 1 select * from test2 where a2 > 15 settings allow_experimental_analyzer = 1; -- buggy, analyzer does not pick up index i
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02911_support_alias_column_in_indices.sql:32]: https://github.com/ClickHouse/ClickHouse/issues/57333
